solidity/test/vault/DonationVault.test.ts

```diff
@@ -0,0 +1,283 @@
+
+  const Bank = await ethers.getContractFactory<Bank__factory>("Bank")
+  const bank = await Bank.deploy()
+  await bank.deployed()
+
+  await bank.connect(deployer).updateBridge(bridge.address)
+
+  const DonationVault = await ethers.getContractFactory<DonationVault__factory>(
+    "DonationVault"
+  )
+  const vault = await DonationVault.deploy(bank.address)
+  await vault.deployed()
+
+  return {
+    bridge,
+    account1,
+    account2,
+    bank,
+    vault,
+  }
+}
+
+describe("DonationVault", () => {
+  let bridge: SignerWithAddress
+  let account1: SignerWithAddress
+  let account2: SignerWithAddress
+  let bank: Bank
+  let vault: DonationVault
+
+  before(async () => {
+    // eslint-disable-next-line @typescript-eslint/no-extra-semi
+    ;({ bridge, account1, account2, bank, vault } = await waffle.loadFixture(
+      fixture
+    ))
+  })
+
+  describe("constructor", () => {
+    context("when called with a 0-address bank", () => {
+      it("should revert", async () => {
+        const DonationVault = await ethers.getContractFactory("DonationVault")
+        await expect(
+          DonationVault.deploy(ethers.constants.AddressZero)
+        ).to.be.revertedWith("Bank can not be the zero address")
+      })
+    })
+
+    context("when called with correct parameters", () => {
+      it("should set the Bank field", async () => {
+        expect(await vault.bank()).to.equal(bank.address)
+      })
+    })
+  })
+
+  describe("donate", () => {
+    context("when caller has not enough balance in the bank", () => {
+      before(async () => {
+        await createSnapshot()
+
+        await bank.connect(bridge).increaseBalance(account1.address, 999)
+      })
+
+      after(async () => {
+        await restoreSnapshot()
+      })
+
+      it("should revert", async () => {
+        await expect(vault.connect(account1).donate(1000)).to.be.revertedWith(
+          "Amount exceeds balance in the bank"
+        )
+      })
+    })
+
+    context(
+      "when vault does not have enough allowance for caller's balance",
+      () => {
+        before(async () => {
+          await createSnapshot()
+
+          await bank.connect(bridge).increaseBalance(account1.address, 1000)
+          await bank
+            .connect(account1)
+            .increaseBalanceAllowance(vault.address, 999)
+        })
+
+        after(async () => {
+          await restoreSnapshot()
+        })
+
+        it("should revert", async () => {
+          await expect(vault.connect(account1).donate(1000)).to.be.revertedWith(
+            "Transfer amount exceeds allowance"
+          )
+        })
+      }
+    )
+
+    context("when called with correct parameters", () => {
+      let tx: ContractTransaction
+
+      before(async () => {
+        await createSnapshot()
+
+        await bank.connect(bridge).increaseBalance(account1.address, 1000)
+        await bank
+          .connect(account1)
+          .increaseBalanceAllowance(vault.address, 1000)
```

Comment:
Can we `increaseBalance` and `increaseBalanceAllowance` for a bit more than `1000` to confirm the amount passed as `donate` parameter is taken into account?

Reply:
See 52fc175